File: Shareable/Shareable/Shareable.cs
using System.Text;
#nullable enable
namespace Shareable
{
    public abstract class Shareable<T>
    {
        public readonly int? Length;
        protected Shareable(int? c = null) { Length = c; }
        public abstract Bookmark<T>? First();
        public T[]? ToArray()
        {
            if (Length == null)
                return null;
            var r = new T[Length.Value];
            for (var b = First(); b != null; b = b.Next())
                r[b.Position] = b.Value;
            return r;
        }
    }
    public abstract class Bookmark<T>
    {
        public readonly int Position;
        protected Bookmark(int p) { Position = p; }
        public abstract Bookmark<T>? Next();
        public abstract T Value { get; }
        public virtual void Append(StringBuilder sb) { }
    }
}

File: 0.0/ShareableJ/src/org/shareabledata/Shareable.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package org.shareabledata;

import java.lang.reflect.Array;

/**
 *
 * @author 66668214
 */
public abstract class Shareable<T> {

    public static boolean HasLength = true;
    public final int Length;

    protected Shareable(int c) {
        Length = c;
    }

    public abstract Bookmark<T> First();

    public T[] ToArray(Class<T> c) {
        T[] r = null;
        r = (T[]) Array.newInstance(c, Length);
        for (Bookmark<T> b = First(); b != null; b = b.Next()) {
            r[b.Position] = b.getValue();
        }
        return r;
    }
}
